subsys/audio_module/audio_module.c

```
@@ -0,0 +1,1065 @@
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_input(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+	struct audio_data block;
+	void *data;
+
+	if (handle == NULL) {
+		LOG_DBG("Module task has NULL handle");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	while (1) {
+		data = NULL;
+
+		if (handle->description->functions->data_process != NULL) {
+			/* Get a new output buffer */
+			ret = k_mem_slab_alloc(handle->thread.data_slab, (void **)&data, K_FOREVER);
+			if (ret) {
+				LOG_DBG("No free data for module %s, ret %d", handle->name, ret);
+				continue;
+			}
+
+			/* Configure new audio block */
+			block.data = data;
+			block.data_size = handle->thread.data_size;
+
+			/* Process the input audio block */
+			ret = handle->description->functions->data_process(
+				(struct audio_module_handle_private *)handle, NULL, &block);
+			if (ret) {
+				k_mem_slab_free(handle->thread.data_slab, (void **)(&data));
+
+				LOG_DBG("Data process error in module %s, ret %d", handle->name,
+					ret);
+				continue;
+			}
+
+			/* Send input audio block to next module(s) */
+			send_to_connected(handle, &block);
+		} else {
+			LOG_DBG("No process function for module %s, no input", handle->name);
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs them out of the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_output(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+
+	struct audio_module_message *msg_rx;
+	size_t size;
+
+	if (handle == NULL) {
+		LOG_DBG("Module task has NULL handle");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	while (1) {
+		msg_rx = NULL;
+
+		ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+							&size, K_FOREVER);
+		if (ret) {
```

Since a K_FOREVER is used here (which is fine to keep the thread idle), I don't think it will ever return. The one thing to check for is an error, and in which case it faults, the next line should be a LOG_ERR and it should return the error instead of continuting.

---

Hang over from the command processing version sorry.

---

Yes this should have a wait value, but it will need to be set for the specific module (e.g. for a stream with 10m buffers if it waits for longer than this there has been an underrun, but this changes when the buffer is 7.5ms). This again could be configurable for each module?